File: virtual-world-simulator/src/Entity/Animals/Turtle.cpp
#include "Turtle.h"

Turtle::Turtle(World* world, Point position) :
	Animal{ AssetManager::getAssetManager()->getAsset("turtle"), world, position, 'T', 2, 1 }
{
}

void Turtle::update(std::fstream& logFile)
{
    int chance = Utility::random(1, 4);
	if (chance == 4)
		Animal::update(logFile);
	else
		Entity::update(logFile);
}

bool Turtle::collision(std::fstream& logFile, Entity& entity)
{
	Turtle* turtle = dynamic_cast<Turtle*>(&entity);
	if (turtle != nullptr)
	{
		this->breed<Turtle>(logFile, *turtle);
		return false;
	}

	if(entity.getStrength() < 5)
		return false;

	return Animal::collision(logFile, entity);
}

File: Java/virtual-world-simulator/src/Entities/Animals/Turtle.java
package Entities.Animals;

import App.World;
import Common.Point;
import Entities.Entity;

import javax.swing.*;
import java.util.Random;

public class Turtle extends Animal
{
    public Turtle(World world, Point position)
    {
        super(new ImageIcon("assets/turtle.png"), world, position, 'T', 2, 1);
    }

    public boolean Collision(Entity entity)
    {
        if(entity instanceof Turtle)
        {
            Turtle turtle = new Turtle(this.world, new Point(-1, -1));
            AddChild(turtle);
            return false;
        }

        if(entity.getStrength() < 5)
            return false;

        return super.Collision(entity);
    }

    public void Update()
    {
        Random rnd = new Random();
        int chance = rnd.nextInt(4);
        if(chance == 3)
            super.Update();
        else
            lifespan++;
    }
}
